examples/platforms/simulation/virtual_time/radio-sim.c

```diff
@@ -0,0 +1,1262 @@
+    event.mDataLength  = 1 + aFrame->mLength; // include channel in first byte
+    sRadioTransmitting = true;
+
+    // 4 bytes of preamble + 1 SOF + 1 PHY header @250kbps + 400us of radio state transision time
+    event.mDelay       = ((aFrame->mLength + 6) * 8 * 1000) / 250 + RADIO_STATE_TRANSITION_DELAY_US;
+    sTransmittingUntil = otPlatTimeGet() + event.mDelay;
+
+    memcpy(event.mData, aMessage, event.mDataLength);
+
+    otSimSendEvent(&event);
+}
+
+void platformRadioReceive(otInstance *aInstance, uint8_t *aBuf, uint16_t aBufLength)
+{
+    assert(sizeof(sReceiveMessage) >= aBufLength);
+
+    memcpy(&sReceiveMessage, aBuf, aBufLength);
+
+    sReceiveFrame.mLength = (uint8_t)(aBufLength - 1);
+
+    radioReceive(aInstance);
+}
+
+static otError radioProcessTransmitSecurity(otRadioFrame *aFrame)
+{
+    otError error = OT_ERROR_NONE;
+#if OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2
+    otMacKeyMaterial *key = NULL;
+    uint8_t           keyId;
+
+    otEXPECT(otMacFrameIsSecurityEnabled(aFrame) && otMacFrameIsKeyIdMode1(aFrame) &&
+             !aFrame->mInfo.mTxInfo.mIsSecurityProcessed);
+
+    if (otMacFrameIsAck(aFrame))
+    {
+        keyId = otMacFrameGetKeyId(aFrame);
+
+        otEXPECT_ACTION(keyId != 0, error = OT_ERROR_FAILED);
+
+        if (keyId == sKeyId)
+        {
+            key = &sCurrKey;
+        }
+        else if (keyId == sKeyId - 1)
+        {
+            key = &sPrevKey;
+        }
+        else if (keyId == sKeyId + 1)
+        {
+            key = &sNextKey;
+        }
+        else
+        {
+            error = OT_ERROR_SECURITY;
+            otEXPECT(false);
+        }
+    }
+    else
+    {
+        key   = &sCurrKey;
+        keyId = sKeyId;
+    }
+
+    aFrame->mInfo.mTxInfo.mAesKey = key;
+
+    if (!aFrame->mInfo.mTxInfo.mIsHeaderUpdated)
+    {
+        otMacFrameSetKeyId(aFrame, keyId);
+        otMacFrameSetFrameCounter(aFrame, sMacFrameCounter++);
+    }
+#else
+    otEXPECT(!aFrame->mInfo.mTxInfo.mIsSecurityProcessed);
+#endif // OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2
+
+    otMacFrameProcessTransmitAesCcm(aFrame, &sExtAddress);
+
+exit:
+    return error;
+}
+
+#if OPENTHREAD_SIMULATION_CCA
+void platformChannelActivity(otInstance *aInstance, uint8_t channel, int8_t value)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    OT_UNUSED_VARIABLE(channel); // For future use of scanning
+
+    if (sCcaPending)
+    {
+        sCcaPending = false;
+        if (value <= sCcaEdThresh)
+        {
+            radioSendMessage(aInstance);
+        }
+        else
+        {
+            platformTransmitReturn(aInstance, &sTransmitFrame, NULL, OT_ERROR_CHANNEL_ACCESS_FAILURE);
+        }
+    }
+}
+
+void simulateCca(otInstance *aInstance, uint8_t channel)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+
+    sCcaPending = true;
+    requestCcaToOtns(channel);
+    otPlatAlarmMicroStartAt(aInstance, otPlatAlarmMicroGetNow(), CCA_DURATION_US);
```

Comment:
IIUC, only the Timer moduel can call `otPlatAlarmMicroStartAt`. No other component can be calling it. 
@jwhui thoughts?

Reply:
This is an important decision. I believe that this piece of code was necessary for timing, or there should be another way to do a similar call.